src/Proto3/Wire/Encode.hs

```diff
@@ -72,25 +73,81 @@ module Proto3.Wire.Encode
     , packedFloats
     , packedDoubles
       -- * Reexports
-    , BB.Builder
-    , BB.toLazyByteString
+    , Builder
+    , builderLength
+    , toLazyByteString
     ) where
 
-import           Data.Bits                   ( (.&.), (.|.), shiftL, shiftR, xor )
-import qualified Data.ByteString             as B
-import qualified Data.ByteString.Builder     as BB
-import qualified Data.ByteString.Lazy        as BL
-import           Data.Int                    ( Int32, Int64 )
-import           Data.Monoid                 ( (<>) )
-import qualified Data.Text.Lazy              as Text.Lazy
-import qualified Data.Text.Lazy.Encoding     as Text.Lazy.Encoding
-import           Data.Word                   ( Word32, Word64, Word8 )
+import           Data.Bits                     ( (.&.), (.|.), shiftL, shiftR, xor )
+import qualified Data.ByteString               as B
+import qualified Data.ByteString.Builder       as BB
+import qualified Data.ByteString.Builder.Extra as BB
+import qualified Data.ByteString.Lazy          as BL
+import           Data.Char                     ( ord )
+import           Data.Int                      ( Int32, Int64 )
+import           Data.Monoid                   ( Sum(..), (<>) )
+import qualified Data.Text.Encoding            as Text.Encoding
+import qualified Data.Text.Lazy                as Text.Lazy
+import qualified Data.Text.Lazy.Encoding       as Text.Lazy.Encoding
+import           Data.Word                     ( Word32, Word64, Word8 )
 import           Proto3.Wire.Types
 
-base128Varint :: Word64 -> BB.Builder
+-- | Like 'BB.Builder', but memoizes the resulting length so
+-- that we can efficiently encode nested embedded messages.
+newtype Builder = Builder { unBuilder :: (Sum Word, BB.Builder) }
+  deriving Monoid
+
+builderLength :: Builder -> Word
+builderLength = getSum . fst . unBuilder
+
+toLazyByteString :: Builder -> BL.ByteString
+toLazyByteString (Builder (Sum len, bb)) =
+    BB.toLazyByteStringWith strat BL.empty bb
+  where
+    -- If the supplied length is accurate then we will perform just
+    -- one allocation (unless that length exceeds 'maxFirstChunk').
+    -- An inaccurate length would indicate a bug in
+    -- one of the primitives that produces a 'Builder'.
+    maxFirstChunk = 134217728  -- 128MiB
```

Comment:
A second, related question: why not set the first chunk size equal to the length?